File: packages/nxuskit-go/auth_oauth.go
//go:build nxuskit

package nxuskit

import (
	"encoding/json"
	"fmt"

	"github.com/nxus-SYSTEMS/nxusKit/packages/nxuskit-go/internal/ffi"
)

// OAuthResult is the result of an OAuth authentication flow.
type OAuthResult struct {
	// Success indicates whether the OAuth flow completed successfully
	Success bool `json:"success"`
	// ProviderID identifies the provider that was authenticated
	ProviderID string `json:"provider_id"`
	// Message is a human-readable status message
	Message string `json:"message"`
	// Error message if the flow failed (nil if successful)
	Error *string `json:"error"`
}

// OAuthStatus is the OAuth authentication status for a provider.
type OAuthStatus struct {
	// Authenticated indicates whether an OAuth credential is stored
	Authenticated bool `json:"authenticated"`
	// ProviderID identifies the provider
	ProviderID string `json:"provider_id"`
	// ExpiresAt is the Unix timestamp when the token expires (nil if unknown)
	ExpiresAt *int64 `json:"expires_at"`
	// Scopes granted by the OAuth token (nil if unknown)
	Scopes []string `json:"scopes"`
}

// OAuthStart initiates an OAuth authentication flow for a provider.
//
// This is a blocking call — it launches a browser, starts a localhost
// callback server, and waits for the authorization code.
//
// Set timeoutSecs to 0 for the default timeout (120 seconds).
func OAuthStart(providerID string, timeoutSecs uint32) (*OAuthResult, error) {
	jsonStr, err := ffi.OAuthStart(providerID, timeoutSecs)
	if err != nil {
		return nil, fmt.Errorf("nxuskit: oauth start failed: %w", err)
	}

	var result OAuthResult
	if err := json.Unmarshal([]byte(jsonStr), &result); err != nil {
		return nil, fmt.Errorf("nxuskit: failed to parse oauth start JSON: %w", err)
	}

	return &result, nil
}

// OAuthStatus checks the OAuth authentication status for a provider.
func OAuthGetStatus(providerID string) (*OAuthStatus, error) {
	jsonStr, err := ffi.OAuthStatus(providerID)
	if err != nil {
		return nil, fmt.Errorf("nxuskit: oauth status failed: %w", err)
	}

	var result OAuthStatus
	if err := json.Unmarshal([]byte(jsonStr), &result); err != nil {
		return nil, fmt.Errorf("nxuskit: failed to parse oauth status JSON: %w", err)
	}

	return &result, nil
}

// OAuthRevoke removes the stored OAuth token for a provider.
//
// Returns nil even if no token was stored.
func OAuthRevoke(providerID string) error {
	return ffi.OAuthRevoke(providerID)
}

File: packages/nxuskit-py/src/nxuskit/auth_oauth.py
"""OAuth authentication for the nxusKit SDK.

Provides browser-based OAuth flow with PKCE and state/CSRF validation
for providers that support it.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import List, Optional

from nxuskit._ffi import ffi, last_error, lib


@dataclass
class OAuthResult:
    """Result of an OAuth authentication flow."""

    success: bool
    """Whether the OAuth flow completed successfully"""
    provider_id: str
    """Provider that was authenticated"""
    message: str
    """Human-readable status message"""
    error: Optional[str]
    """Error message if the flow failed"""


@dataclass
class OAuthStatus:
    """OAuth authentication status for a provider."""

    authenticated: bool
    """Whether an OAuth credential is stored"""
    provider_id: str
    """Provider identifier"""
    expires_at: Optional[int]
    """Unix timestamp when the token expires (None if unknown)"""
    scopes: Optional[List[str]]
    """Scopes granted by the OAuth token (None if unknown)"""


def oauth_start(provider_id: str, timeout_secs: int = 0) -> OAuthResult:
    """Start an OAuth authentication flow for a provider.

    This is a **blocking** call — it launches a browser, starts a localhost
    callback server, and waits for the authorization code.

    Args:
        provider_id: Provider to authenticate (e.g., "azure-openai").
        timeout_secs: Max seconds to wait for callback (0 = default 120s).

    Returns:
        OAuth flow result.

    Raises:
        RuntimeError: If the C ABI call fails.
    """
    c_pid = ffi.new("char[]", provider_id.encode("utf-8"))
    ptr = lib.nxuskit_oauth_start(c_pid, timeout_secs)

    if ptr == ffi.NULL:
        err = last_error() or "unknown error"
        raise RuntimeError(f"nxuskit_oauth_start failed: {err}")

    try:
        raw = json.loads(ffi.string(ptr).decode("utf-8"))
    finally:
        lib.nxuskit_free_string(ptr)

    return OAuthResult(
        success=raw.get("success", False),
        provider_id=raw["provider_id"],
        message=raw["message"],
        error=raw.get("error"),
    )


def oauth_status(provider_id: str) -> OAuthStatus:
    """Check OAuth authentication status for a provider.

    Args:
        provider_id: Provider to check.

    Returns:
        OAuth status including whether authenticated.

    Raises:
        RuntimeError: If the C ABI call fails.
    """
    c_pid = ffi.new("char[]", provider_id.encode("utf-8"))
    ptr = lib.nxuskit_oauth_status(c_pid)

    if ptr == ffi.NULL:
        err = last_error() or "unknown error"
        raise RuntimeError(f"nxuskit_oauth_status failed: {err}")

    try:
        raw = json.loads(ffi.string(ptr).decode("utf-8"))
    finally:
        lib.nxuskit_free_string(ptr)

    return OAuthStatus(
        authenticated=raw["authenticated"],
        provider_id=raw["provider_id"],
        expires_at=raw.get("expires_at"),
        scopes=raw.get("scopes"),
    )


def oauth_revoke(provider_id: str) -> None:
    """Remove the stored OAuth token for a provider.

    Args:
        provider_id: Provider whose OAuth token to remove.

    Raises:
        RuntimeError: If the revocation fails.
    """
    c_pid = ffi.new("char[]", provider_id.encode("utf-8"))
    result = lib.nxuskit_oauth_revoke(c_pid)
    if result < 0:
        err = last_error() or "unknown error"
        raise RuntimeError(f"OAuth revoke failed: {err}")
